Silent Payments: Receiving

Eunovo opened the pull request:
This PR is part of integrating silent payments into Bitcoin Core. Status and tracking for the project is managed in https://github.com/bitcoin/bitcoin/issues/28536

Prerequisite PRs:
- [ ] https://github.com/bitcoin/bitcoin/pull/35301
- [ ] https://github.com/bitcoin/bitcoin/pull/35302

This PR:
- Adds a Silent Payments descriptor implemenation
- Adds a `SilentPaymentsDescriptorScriptPubKeyMan` Impl that is a subclass of `DescriptorScriptPubKeyMan`
- Implements Silent Payments scanning for the wallet
- Updates sending logic to use `silent-payments` destination for change when sending to `silent-payments` destination
- Adds unit and functional tests for silent payments-related functionality

**Follow-ups**
- Silent Payments Label functionality is incomplete in this PR and will be added as a follow-up PR.

josibake committed:
Add "sp" HRP

josibake committed:
wallet: make coin selection silent payment aware

Add a flag to the `CoinControl` object if silent payment destinations
are provided. Before adding the flag, call a function which checks if:

* The wallet has private keys
* The wallet is unlocked

Without both of the above being true, we cannot send to a silent payment
address.

During coin selection, if this flag is set, skip taproot inputs when
script spend data is available. This is based on the assumption that if
a user provides script spend data, they don't have access to the key
path spend. As future improvement, we could instead check to see if we
have access to the key path spend, and only exclude the output when we
don't regardless of whether or not the user provides script spend data.

Also skip UTXOs of type `WITNESS_UNKNOWN`, although it is very unlikely
our wallet would ever try to spend a witness unknown output.

josibake committed:
wallet: add method for retreiving a private key

Add a method for retreiving a private key for a given scriptPubKey.
If the scriptPubKey is a taproot output, tweak the private key with the
merkle root or hash of the public key, if applicable.

josibake committed:
wallet: update TransactionChangeType

If sending to a silent payment destination, the change type should be taproot

josibake committed:
tests: add sending functional tests

josibake committed:
wallet: add `IsInputForSharedSecretDerivation` function

josibake committed:
Add V0SilentPaymentDestination address type

josibake committed:
wallet: disable sending to silent payment address

Have `IsValidDestination` return false for silent payment destinations
and set an error string when decoding a silent payment address.

This prevents anyone from sending to a silent payment address before
sending is implemented in the wallet, but also allows the functions to
be used in the unit testing famework.

josibake committed:
wallet: enable sending to silent payment address

Treat silent payment addresses as valid destination. Also disable using
silent payment addresses with the `addr()` descriptor, as this
descriptor expects an encoding of a scriptPubKey, whereas a silent
payment address consists of instructions on how to generate a
scriptPubKey for the recipient.

Co-authored-by: Oghenovo Usiwoma <37949128+eunovo@users.noreply.github.com>

josibake committed:
common: add bip352.{h,cpp} secp256k1 module

Wrap the silentpayments module from libsecp256k1. This is placed in
common as it is intended to be used by:

  * RPCs: for parsing addresses
  * Wallet: for sending, receiving, spending silent payment outputs
  * Node: for creating silent payment indexes for light clients

josibake committed:
tests: add BIP352 test vectors as unit tests

Use the test vectors to test sending and receiving. A few cases are not
covered here, namely anything that requires testing specific to the
wallet. For example:

* Taproot script path spending is not tested, as that is better tested in
  a wallets coin selection / signing logic
* Re-computing outputs during RBF is not tested, as that is better
  tested in a wallets RBF logic

The unit tests are written in such a way that adding new test cases is
as easy as updating the JSON file

josibake committed:
wallet: get serialized size for `V0SilentPayments`

BIP352 v0 specifies that a silent payment output is a taproot output.
Taproot scriptPubKeys are a fixed size, so when calculating the
serialized size for a CRecipient with a V0SilentPayments destination,
use WitnessV1Taproot for the serialized txout size.

josibake committed:
wallet: add `CreateSilentPaymentOutputs` function

`CreateSilentPaymentsOutputs` gets the correct private keys, adds them
together, groups the silent payment destinations and then generates the
taproot script pubkeys. These are then passed back to
CreateTransactionInternal, which uses these scriptPubKeys to update
vecSend before adding them to the transaction outputs.

josibake committed:
crypto: add read-only method to KeyPair

Add a method for passing a KeyPair object to secp256k1 functions expecting a secp256k1_keypair.
This allows for passing a KeyPair directly to a secp256k1 function without needing to create a
temporary secp256k1_keypair object.

Eunovo committed:
wallet: generate sp() descriptor

The scan key will be derived while the descriptor is parsed, so
we must pass the master xpriv into `GenerateWalletDescriptor`.

The spend key is expanded later so it is specified
as an xpub and path instead.

Eunovo committed:
wallet: create SilentPaymentDescriptorScriptPubKeyMan

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo committed:
wallet/tests: Enable Silent Payments in wallet unit tests

Eunovo committed:
wallet: import silent payments descriptor

Eunovo committed:
tests: add silent payments receiving functional tests

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo committed:
update addressbook with found silent-payment outputs

traditionally, the receiving scripts are known to the core wallet
because they are added to the addressbook at the time they are requested.

in silent payments, the outputs are not known until found. its important, however,
to have the found scripts in the addressbook so all of the change accounting can be down
properly.

this commits adds found outputs to the address book if they are not change. the way
change is determined is a bit hacky in that we just check if the found output was
found with a label (since this is the only label currently implemented). in the future,
we should specifically check for a change specific label.

josibake committed:
use silent payments change when sending

The wallet tries to match change OutputTpe to other outputs to preserve privacy.
Use silent payment change address when sending coins to:
- any silent payment address
- any taproot address

Eunovo committed:
wallet/rpc: add create silent-payments wallet option

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo committed:
key: add ComputeBIP352Key() method

This method makes it possible to tweak silent payment spend keys in later commits

Eunovo committed:
wallet/rpc: add sp-address support to getaddressinfo

Sjors commented:
Another thing I noticed while testing using the `send` RPC, in a wallet with _only_ an `sp()` descriptor, is that it insists on having a bech32 descriptor for change:

> Transaction needs a change address, but we can't generate it. Error: No bech32 addresses available.

It seems safer, for backups, to send change to the same silent payment address.

Wallet created by:

```sh
bitcoin rpc createwallet Silent blank=true silent_payments=true
bitcoin rpc importdescriptors '[{"desc": "sp(xprv...,xprv...)",  "timestamp": "now", "active": true]
```

Eunovo commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -358,6 +358,7 @@ static RPCHelpMan createwallet()
             {"descriptors", RPCArg::Type::BOOL, RPCArg::Default{true}, "If set, must be \"true\""},
             {"load_on_startup", RPCArg::Type::BOOL, RPCArg::Optional::OMITTED, "Save wallet name to persistent settings and load on startup. True to add wallet to startup list, false to remove, null to leave unchanged."},
             {"external_signer", RPCArg::Type::BOOL, RPCArg::Default{false}, "Use an external signer such as a hardware wallet. Requires -signer to be configured. Wallet creation will fail if keys cannot be fetched. Requires disable_private_keys and descriptors set to true."},
+            {"silent_payments", RPCArg::Type::BOOL, RPCArg::Default{false}, "Enable creating and receiving with a silent payments address. Will disable fast rescans with block filters"},
```
Thanks @Sjors. I'll fix it

josibake replied:
We can disable the fast rescan for now just to keep this PR focused, but in theory we should be able to use the fast rescan regardless of whether or not the wallet is a silent payments wallet. Perhaps we can leave a comment to enable fast rescans in a follow up?

Eunovo replied:
IIUC The silent payments wallet needs to scan entire blocks since it doesn't know ahead of time which `scriptPubKeys` to use in a GCS filter. Since we can't skip blocks, fast rescans shouldn't provide any advantages here.

Eunovo replied:
> This makes sense. Is it documented anywhere so we don't forget by the next review round? :-)

I added a separate commit that disables fast rescan and I added an explanation to the commit message. Will push soon.

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    // Populate m_map_label_tweaks
+    auto change_label_data = bip352::CreateLabelTweak(provider.sp_keys.first, 0);
+    m_map_label_tweaks.insert(change_label_data);
+    for (int i = 1; i < m_wallet_descriptor.next_index; i++) {
+        // Add the other generated labelled destinations
+        m_map_label_tweaks.insert(bip352::CreateLabelTweak(provider.sp_keys.first, i));
+    }
```
I'd remove this (and any other code) related to non-change labels from this PR. Ideally, implement just enough for labels to get change working, and then we can fully implement labels in a follow-up PR.

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    if (m_wallet_descriptor.descriptor->GetOutputType() != OutputType::SILENT_PAYMENTS) {
+        throw std::runtime_error(std::string(__func__) + ": descriptor is not a Silent Payment Descriptor");
+    }
+    const auto provider{GetSPProvider()};
```
It feels strange to call the provider just to get the scan key, feels like another area we can simplify with a custom scan key type.

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    return dest;
+}
+
+V0SilentPaymentDestination SilentPaymentDescriptorScriptPubKeyMan::GetLabeledDestination(uint64_t index)
+{
+    AssertLockHeld(cs_desc_man);
+
+    const auto provider{GetSPProvider()};
+    V0SilentPaymentDestination main_dest;
+    main_dest.m_scan_pubkey = provider.sp_keys.first.GetPubKey();
+    main_dest.m_spend_pubkey = provider.sp_keys.second;
+    // Generate label tweak
+    auto label_data = bip352::CreateLabelTweak(provider.sp_keys.first, index);
+    m_map_label_tweaks.insert(label_data);
+    V0SilentPaymentDestination label_dest = bip352::GenerateSilentPaymentLabeledAddress(main_dest, label_data.second);
+    return label_dest;
+}
+
+util::Result<CTxDestination> SilentPaymentDescriptorScriptPubKeyMan::GetNewLabeledDestination(uint64_t& index)
+{
+    LOCK(cs_desc_man);
+    auto dest = GetLabeledDestination(m_wallet_descriptor.next_index);
+    index = m_wallet_descriptor.next_index; // Return the index for this destination
+    m_wallet_descriptor.next_index++;
+    WalletBatch(m_storage.GetDatabase()).WriteDescriptor(GetID(), m_wallet_descriptor);
+    return CTxDestination{dest};
+}
```
Same comment regarding moving labels related code to a follow-up PR.

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    return CTxDestination{op_dest};
+}
+
+bool SilentPaymentDescriptorScriptPubKeyMan::TopUp(const uint256& tweak)
```
The name TopUp here makes sense in that we are trying to match how the rest of the wallet does it, but @Sjors pointed out this can be confusing for people who think of TopUp in terms of the gap limit / key pool refresh context, which we aren't really doing here. We could just have one function called `AddTweak` and then use a boolean for writing to the database, e.g., `AddTweak(uint256& tweak, store_in_db=false)`

Eunovo replied:
Updated

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    return res;
+}
+
+bool SilentPaymentDescriptorScriptPubKeyMan::TopUpWithDB(WalletBatch& batch, const uint256& tweak)
```
Could be `AddTweakWithDB` (based on my previous comment).

Eunovo replied:
Updated

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1635,4 +1635,212 @@ bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescript
+    return batch.WriteSilentPaymentsTweak(GetID(), tweak);
+}
+
+FlatSigningProvider SilentPaymentDescriptorScriptPubKeyMan::GetSPProvider() const
+{
+    AssertLockHeld(cs_desc_man);
+    FlatSigningProvider provider;
+    provider.keys = GetKeys();
+    FlatSigningProvider tmp;
+    m_wallet_descriptor.descriptor->ExpandPrivate(0, provider, tmp);
+    return tmp;
+}
```
Can likely be removed if we have custom methods for storing scan/spend pubkey data, or custom methods for getting them.

josibake commented on src/interfaces/chain.h:
```diff
@@ -62,6 +62,12 @@ class FoundBlock
     //! Read block data from disk. If the block exists but doesn't have data
     //! (for example due to pruning), the CBlock variable will be set to null.
     FoundBlock& data(CBlock& data) { m_data = &data; return *this; }
+    //! Read block's undo data from disk.
```
This commit is quite long, making it difficult to focus on the main thing which is the scanning logic. I'd suggest breaking out smaller commits for just the mempool related changes, and just the undoblock related changes. Possibly more, but will need to dig into the commit a bit more.

Eunovo replied:
I broke this into 3 commits.

josibake commented on src/wallet/wallet.cpp:
```diff
@@ -1464,9 +1466,30 @@ void CWallet::blockConnected(ChainstateRole role, const interfaces::BlockInfo& b
     // Scan block
     bool wallet_updated = false;
     for (size_t index = 0; index < block.data->vtx.size(); index++) {
+        CTransactionRef tx{block.data->vtx.at(index)};
```
Do we need the `CTransactionRef` here, as opposed to just operating on `block.data->vtx` directly?

Eunovo replied:
We only use the reference once, so it's not needed. I'll take this out as I retouch.

Eunovo replied:
I ended up leaving the reference and used it in 2 other locations. The code looks slightly cleaner this way.

josibake reviewed (commented):
Did a big overview with @Sjors , leaving the notes from our discussion as a review. In general, I think we should investigate using a custom type for the scan key since a lot of the changes seem to be hacking around the fact we represent the scan key as a private key, but then often need to use it as not a private key.

Still in the process of reviewing, but leaving the initial notes for now

Eunovo commented:
Thanks for the reviews @josibake and @Sjors . I am currently travelling, and I will get to them later in the week.

josibake commented on src/script/descriptor.cpp:
```diff
@@ -2120,6 +2120,10 @@ std::vector<std::unique_ptr<DescriptorImpl>> ParseScript(uint32_t& key_exp_index
             error = "Address is not valid";
             return {};
         }
+        if (std::holds_alternative<V0SilentPaymentDestination>(dest)) {
```
Cherry-picked this commit into https://github.com/bitcoin/bitcoin/pull/28201/commits/b94934e089e70b478ea6aa535098586fedb5a12d, in #28201. I added it to the commit where we enable silent payment addresses as valid addresses, at the end of the sending PR.

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -1422,16 +1422,78 @@ class RawTRDescriptor final : public DescriptorImpl
+        m_pubkey_args.at(1)->GetPrivKey(0, provider, tmp);
+
+        assert(spend_pubkey.has_value());
+        out.sp_keys = std::make_pair(m_scan_key, *spend_pubkey);
+
+        auto it{tmp.keys.find(spend_pubkey->GetID())};
+        if (it != tmp.keys.end()) out.keys.emplace(spend_pubkey->GetID(), it->second);
```
Without `sp_keys`, the wallet can't identify which keys are scan and spend keys from the signing provider. Do you have an alternative approach to this?

Eunovo committed:
Add OutputType::SILENT_PAYMENTS

Eunovo committed:
descriptor: implement sp(scan, spend) desc

sp() expects a scan private key and a spend key which could be a private key
or a public key.
The scan key is really a "scan entropy" and will not be treated as an actual
private key.

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -1422,16 +1422,78 @@ class RawTRDescriptor final : public DescriptorImpl
+
+    std::optional<OutputType> GetOutputType() const override { return OutputType::SILENT_PAYMENTS; }
+
+    bool IsSingleType() const final { return true; }
```
Done

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -1422,16 +1422,78 @@ class RawTRDescriptor final : public DescriptorImpl
-    P2SH,    //!< Inside sh() (script becomes P2SH redeemScript)
-    P2WPKH,  //!< Inside wpkh() (no script, pubkey only)
-    P2WSH,   //!< Inside wsh() (script becomes v0 witness script)
-    P2TR,    //!< Inside tr() (either internal key, or BIP342 script leaf)
```
Fixed.

Eunovo commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -481,7 +481,7 @@ BOOST_FIXTURE_TEST_CASE(BasicOutputTypesTest, ListCoinsTest)
 
     for (const auto& out_type : OUTPUT_TYPES) {
-        if (out_type == OutputType::UNKNOWN) continue;
+        if (out_type == OutputType::UNKNOWN || out_type == OutputType::SILENT_PAYMENTS) continue;
```
Done

Eunovo commented on src/wallet/wallet.cpp:
```diff
@@ -3523,6 +3523,7 @@ void CWallet::SetupDescriptorScriptPubKeyMans(WalletBatch& batch, const CExtKey&
     AssertLockHeld(cs_wallet);
     for (bool internal : {false, true}) {
         for (OutputType t : OUTPUT_TYPES) {
+            if (t == OutputType::SILENT_PAYMENTS) continue;
```
Done

Eunovo commented on src/outputtype.cpp:
```diff
@@ -18,6 +18,7 @@ static const std::string OUTPUT_TYPE_STRING_LEGACY = "legacy";
 static const std::string OUTPUT_TYPE_STRING_P2SH_SEGWIT = "p2sh-segwit";
 static const std::string OUTPUT_TYPE_STRING_BECH32 = "bech32";
 static const std::string OUTPUT_TYPE_STRING_BECH32M = "bech32m";
+static const std::string OUTPUT_TYPE_STRING_SILENT_PAYMENTS = "silent-payments";
```
Done

Eunovo commented on src/key.cpp:
```diff
@@ -166,6 +166,13 @@ void CKey::MakeNewKey(bool fCompressedIn) {
     fCompressed = fCompressedIn;
 }
 
+bool CKey::TweakAdd(const unsigned char* tweak32)
```
Maybe they deserve their own PR?

Eunovo commented on src/wallet/walletutil.cpp:
```diff
@@ -73,28 +75,41 @@ WalletDescriptor GenerateWalletDescriptor(const CExtPubKey& master_key, const Ou
+        // The actual scan and spend keys will be derived from these XPRIVs
+        // The actual XPRIV and the spend key will not be retained in the descriptor,
+        // but the scan key will be.
+        desc_str = "sp(" + xpriv + "/352h/0h/0h/1h/0," + xpriv + "/352h/0h/0h/0h/0)";
```
Don't we need to derive the scan and spend keys?

Eunovo commented on src/wallet/walletutil.cpp:
```diff
@@ -73,28 +75,41 @@ WalletDescriptor GenerateWalletDescriptor(const CExtPubKey& master_key, const Ou
     }
     } // no default case, so the compiler can warn about missing cases
-    assert(!desc_prefix.empty());
+    assert(!desc_str.empty() || !desc_prefix.empty());
```
In the `OutputType::SILENT_PAYMENTS` case, I set the `desc_str` not the `desc_prefix`; I have to modify the `assert` to prevent a failure.

Eunovo committed:
notifications: Add spent_coins to tx added notification

The spent_coins data will be used to scan for silent payments in later commits

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo committed:
interfaces: Load block undo data

This undo data will be used to scan for silent payments in later commits

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo committed:
wallet: scan for silent payments

Co-authored-by: Ava Chow <github@achow101.com>

Eunovo commented:
Added @achow101 as co-author on commits with code/ideas taken from  https://github.com/bitcoin/bitcoin/pull/28453

Eunovo committed:
wallet: disable fast rescan for silent_payments wallet

silent_payments wallet needs to scan entire blocks since it doesn't know ahead of time which scriptPubKeys to use in it's fast rescan filter.

Eunovo commented:
> Another thing I noticed while testing using the `send` RPC, in a wallet with _only_ an `sp()` descriptor, is that it insists on having a bech32 descriptor for change:
> 
> > Transaction needs a change address, but we can't generate it. Error: No bech32 addresses available.

In https://github.com/bitcoin/bitcoin/pull/32966/commits/ed7f894fc0c1ed3a84f4af73ece619698cee11c0 I modify the change `OutputType` determination logic to try to use silent payments if `bech32m` addresses are not available.

> It seems safer, for backups, to send change to the same silent payment address.

That may be true, but the wallet tries to match change output to other outputs to preserve privacy.

josibake commented:
Looks like the most recent change is causing a failure in the unit tests:

> 2025-07-22T13:09:52.932304Z (mocktime: 2020-08-31T15:34:12Z) [test] [txmempool.cpp:699] [void CTxMemPool::check(const CCoinsViewCache &, int64_t) const] [mempwallet/spend.cpp:1343 CreateTransactionInternal: Assertion `!change_script.empty()' failed.

Eunovo committed:
wallet/spend: Filter out coins not eligible to pay sp change dest

Eunovo committed:
wallet: retry CreateTransaction without silent_payments change

`CreateTransaction` fails if change_type is set to OutputType::SILENT_PAYMENTS
and there are no eligible inputs for the transaction. This commit modifies
`CreateTransaction` logic to make a second attempt to
create a tx without OutputType::SILENT_PAYMENTS change_type

Eunovo committed:
wallet: prefer silent_payments over Bech32m as fallback change type

This allows wallets with only sp descs to use sp change destination for transactions.

josibake commented:
> The current failing CI

Should be fixed in my (just now) pushed changes for the base and sending PRs.

Sjors commented on test/functional/wallet_silentpayments_receiving.py:
```diff
@@ -147,7 +147,30 @@ def do_nothing():
+        res = wallet.importdescriptors([{
+            "desc": descriptor,
+            "active": True,
+            "next_index": 0,
```
5be304dfa50bcfd3b0a92eff52bac3d8023ce523 nit: no need to set `next_index` during import

Eunovo replied:
Done

macgyver13 commented on src/script/descriptor.cpp:
```diff
@@ -1772,6 +1899,62 @@ std::vector<std::unique_ptr<DescriptorImpl>> ParseScript(uint32_t& key_exp_index
+            }
+        }
+
+        // Remove all scan private keys to allow
+        // importation of sp descriptors into watch-only wallets
+        for (auto keyId : scan_keys_rm) {
+            out.keys.erase(keyId);
+        }
```
I think this change is breaking the ability to import sp descriptor with private keys defined for scan and spend
This is the warning I receive when using importdescriptors:
"Not all private keys provided. Some wallet functionality may return unexpected errors"

Eunovo replied:
@macgyver13  Did you use the same private key for scan and spend?

macgyver13 replied:
This is the format I used to import scan and spend private keys:
sp([fingerprint/352'/1'/0'/1'/0]WIF(private scan key),[fingerprint/352'/1'/0'/0'/0]WIF(private spend key)#checksum

macgyver13 commented on test/functional/wallet_silentpayments_receiving.py:
```diff
@@ -0,0 +1,332 @@
+        self.nodes[0].createwallet(wallet_name="import_blank", blank=True, silent_payments=True)
+        wallet = self.nodes[0].get_wallet_rpc("import_blank")
+
+        descriptor = "sp([eea23daf/352h/0h/0h/1h/0]cRCGWnoELVHTr9oZWz1TUp7jmdBJPS3Kx8UCaHvfpxL6KiVSHH1A,[eea23daf/352h/0h/0h/0h/0]cUxxbQ67tepsEn3AUKSCWAKrvLwP6MPk55DAptcqxqDiXN6yKkNW)#q42ue5dg"
```
Shouldn't the coin_type be different for regtest/signet? /352h/1h instead of /352h/0h

Sjors replied:
That's definitely a good idea. Some hardware wallets may even refuse to sign if the coin type derivation is wrong, e.g. to prevent a testnet app from moving mainnet coins.

Eunovo replied:
I've changed this to use to `/352h/1h`. I also edited `GenerateWalletDescriptor` to use `/352h/1h` for test chain.

macgyver13 reviewed (commented):
working on backup / restore testing I noticed a few issues.

Eunovo committed:
wallet/gui: add silent_payments option to createwalletdialog

Eunovo committed:
wallet/gui: add silent-payments to address type dropdown

Eunovo committed:
descriptors: impl ScankeyPubkeyProvider

While the scankey looks like a private key, and is represented as a `CKey` in the code,
it is actually much closer to a public key and it needs to be treated as a public key,
so it can be added to watch-only wallets.

josibake commented:
Rebased the base and sending PRs on master and fixed the static assert failure. Rebasing this on the sending PR should fix the CI failure.

josibake commented on src/script/descriptor.cpp:
```diff
@@ -1026,16 +1026,23 @@ class DescriptorImpl : public Descriptor
 
     // NOLINTNEXTLINE(misc-no-recursion)
-    void GetPubKeys(std::set<CPubKey>& pubkeys, std::set<CExtPubKey>& ext_pubs) const override
+    void GetPubKeys(std::set<CPubKey>& pubkeys, std::set<CExtPubKey>& ext_pubs, std::vector<CKey>& scankeys) const override
```
Another idea: add a new method to the descriptor class , `GetPubKeyProviders` ? This avoids needing to subclass just for silent payments, and may prove to be a generally useful method. This way, we can return the ScankeyPubkeyProvider and access the scan "private" key directly from that.

josibake reviewed (commented):
Looks much better now that you added `ScankeyPubkeyProvider`. We could even just call this a `SilentPaymentsPubKeyProvider`, which is perhaps a bit less confusing for reviewers who are less familiar with the details of how silent payments work. Reviewed up until the `SilentPaymentsSPKMan` commit and onward.

Eunovo committed:
tests: silent-payments wallet coin control

Co-authored-by: macgyver13 <4712150+macgyver13@users.noreply.github.com>

Eunovo committed:
descriptor: add SPDescriptor public interface

Create an SPDescriptor interface that will be used by the
`SilentPaymentDescriptorScriptPubKeyMan`.

The interface only needs to provide a way to get the scan key,
the spend key can be obtained through the normal Descriptor methods.

While the scan key is represented as a `CKey`, it will not be used
in the signing process. This creates an argument against putting this
in a signing provider.

Eunovo commented:
Some Major Updates:
- The `SilentPaymentDescriptorScriptPubKeyMan` now obtains through a `GetScanKey` method, only available on Silent Payment descriptors
- The spend key is no longer derived during descriptor parsing; it now has to be "expanded" the same way other descriptors are expanded
- `LoadKeysAndChangeLabel` has been replaced with an overridden `TopUpWithDB` method. The wallet calls `TopUp` when it wants the SPKMans to derive and load new keys and scripts.

josibake reviewed (commented):
Looking good! Really nice to see you were able to remove a lot of custom code related to how we handle the scan and spend keys. Left some comments/questions below.

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -1693,9 +1692,18 @@ class SpDescriptor final : public DescriptorImpl
         return true;
     }
 
+    CKey GetScanKey() const override
+    {
+        FlatSigningProvider dummy, out;
+        m_scan_pubkey->GetPrivKey(0, dummy, out);
+        // The ScankeyPubkeyProvider will always contain the scan key
+        assert(out.keys.size() == 1);
+        return out.keys.begin()->second;
+    }
```
Maybe we can use a namespace?

Eunovo commented on src/wallet/spend.cpp:
```diff
@@ -1307,25 +1306,42 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
                 sp_dests[i] = *sp;
             }
         }
+        size_t sp_change_index{sp_dests.size() > 0 ? sp_dests.rbegin()->first + 1 : 0};
```
We need it to extract the `change_dest` from the results of `CreateSilentPaymentOutputs`.

josibake commented on src/key.h:
```diff
@@ -128,6 +128,8 @@ class CKey
     //! Generate a new private key using a cryptographic PRNG.
     void MakeNewKey(bool fCompressed);
 
+    bool TweakAdd(const unsigned char* tweak32);
```
FYI, I have a WIP commit that implements a `SilentPaymentsSign` function, I'll push the commit later today. Its a bit more code than what we have with `CKey::TweakAdd`, but I do like that its silent payments specific and can't be used as a footgun.

Eunovo committed:
bip352: Make SilentPaymentOuput serializable

In future commits, a Silent Payment enabled wallet will write found SilentPaymentOutputs to the DB.
Although the XOnlyPubKey can be derived from adding the Spend pubkey and the tweak,
we opt  to save the the found XOnlyPubKey because that removes the need for a CPubkey::TweakAdd() in the codebase

Sjors commented:
@Eunovo when I call `setwalletflag` it fails, saying the flag is immutable. I'm fine with either solution:
1. Make it mutable; or
2. Set it when importing an sp descriptor

PedroRosalba commented:
hey guys what is the status of this - is it stopped? in case it's ongoing, are there any blockers?

Eunovo commented:
> hey guys what is the status of this - is it stopped? in case it's ongoing, are there any blockers?

The major blocker is https://github.com/bitcoin-core/secp256k1/issues/1799. We need this issue to be resolved so we can get the BIP352 primitives into a libsecp256k1 release

Eunovo committed:
Squashed 'src/secp256k1/' changes from 7262adb4b4..14be7f2abe

14be7f2abe silentpayments: skip slow benchmarks for low iters count (<= 2)
ea46d6502e docs: update README
8587aeb029 ci: enable silentpayments module
1375d38a32 tests: add sha256 tag test
eb3748dab1 tests: add constant time tests
e1674f389a tests: add BIP-352 test vectors
9163035ab7 silentpayments: optimize scanning by using batch inversion
dc36ff6ec3 silentpayments: add benchmarks for scanning
b7f09f3eb5 silentpayments: add examples/silentpayments.c
460565c662 silentpayments: respect per-group recipients protocol limit (K_max=2323)
72658fc85e silentpayments: receiving
c7a64d0cd8 silentpayments: recipient label support
be56a66992 silentpayments: sending
7bfae739e3 build: add skeleton for new silentpayments (BIP352) module
b11340b3ce Merge bitcoin-core/secp256k1#1849: musig: always clear out secret key in `secp256k1_musig_nonce_gen_counter`
8479eafa57 musig: always clear out secret key in `secp256k1_musig_nonce_gen_counter`
c1a9e4fe64 Merge bitcoin-core/secp256k1#1848: ci: Bump GCC snapshot major version to 17
3cca6451a2 ci: Bump GCC snapshot major version to 17
ea174fe045 Merge bitcoin-core/secp256k1#1846: ci: Replace `ilammy/msvc-dev-cmd` with manual MSVC setup
285cb788e9 ci: Replace `ilammy/msvc-dev-cmd` with manual MSVC setup

git-subtree-dir: src/secp256k1
git-subtree-split: 14be7f2abe4797c1a58249cb901cc6f02972cb4d

Eunovo committed:
Merge commit 'bb1bce36cf8b9292a65f0f03ab855e7133357b9f' into new-sp-base

rkrux commented on test/functional/wallet_silentpayments_receiving.py:
```diff
@@ -0,0 +1,291 @@
+        self.log.info("Check bumpfee on tx with sp inputs")
+
+        address = self.def_wallet.getnewaddress()
+        txid = wallet.sendtoaddress(address, 10, replaceable=True)
```
In ffa0e15a8359cd1c71e74f21f157c0d6f28eedb6 "tests: add silent payments receiving functional tests"

> replaceable=True

Similar to the suggestion in the sending PR, this should not be required.

rkrux commented on test/functional/wallet_silentpayments_receiving.py:
```diff
@@ -0,0 +1,291 @@
+        self.disconnect_nodes(0, 1)
+        txid = self.def_wallet.sendtoaddress(addr, send_amount)
+        raw_tx = self.nodes[0].getrawtransaction(txid)
+        def do_nothing():
+            pass
+        self.nodes[1].sendrawtransaction(raw_tx)
+        self.generate(self.nodes[1], 1, sync_fun=do_nothing)
```
In ffa0e15 "tests: add silent payments receiving functional tests"

It looks like do_nothing() is a duplicate of no_op() that's present in the testing framework?

rkrux commented on test/functional/wallet_silentpayments_receiving.py:
```diff
@@ -0,0 +1,291 @@
+        self.nodes[0].createwallet(wallet_name="import_same_scan_spend", blank=True, silent_payments=True)
+        wallet = self.nodes[0].get_wallet_rpc("import_same_scan_spend")
+
+        descriptor = "sp([eea23daf/352h/1h/0h/0h/0]cUxxbQ67tepsEn3AUKSCWAKrvLwP6MPk55DAptcqxqDiXN6yKkNW,[eea23daf/352h/1h/0h/0h/0]cUxxbQ67tepsEn3AUKSCWAKrvLwP6MPk55DAptcqxqDiXN6yKkNW)#vktsamkr"
```
In ffa0e15 "tests: add silent payments receiving functional tests"

Let's avoid hardcoding the descriptor checksum and instead use the `descsum_create` utility function.

rkrux reviewed (commented):
Partial review of the testing commit ffa0e15 "tests: add silent payments receiving functional tests".

Since this PR is stacked over two open PRs at the moment, this feedback can be addressed when it's rebased. I shared because these caught my attention while giving a cursory glance at the test.
